File: api/routes/diagnostics.py
"""Diagnose & solve problems — Reliability / Availability routes.

Read-only best-practice checks over the configured Azure resources. Unlike the
`/api/monitor/*` surface (which degrades open to an empty payload), this surface
reports a fetch failure / permission denial as an `indeterminate` finding, so a
"could not check" never masquerades as "no problems found".

Responsibility: HTTP validation + auth + response shaping for
    `GET /api/diagnostics/{category}`. Delegate all fetch + rule logic to
    `api.services.diagnostics`.
Edit boundaries: Keep HTTP concerns here; no Azure SDK, no rule logic.
Key entry points: `diagnostic_report`.
Risky contracts: Every route enforces `require_caller`. This is a plain
    request/response GET (no SSE), so §12a Rule 5 does not apply. Findings are
    already sanitised by the engine.
Validation: `uv run pytest -q api/tests/test_diagnostics_route.py
    api/tests/test_route_contracts.py`.
"""

from __future__ import annotations

import logging
from typing import Any

from fastapi import APIRouter, Depends, HTTPException, Query

from api.auth import CallerIdentity, require_caller
from api.routes.monitor.common import _sub_default
from api.services.diagnostics.engine import run_diagnostic, supported_categories
from api.services.diagnostics.snapshot import DiagnosticTarget

LOGGER = logging.getLogger(__name__)

router = APIRouter(prefix="/api/diagnostics", tags=["diagnostics"])


@router.get("/{category}")
def diagnostic_report(
    category: str,
    subscription_id: str = Query(default=""),
    workload_resource_group: str = Query(default=""),
    acr_resource_group: str = Query(default=""),
    acr_name: str = Query(default=""),
    storage_account_name: str = Query(default=""),
    region: str = Query(default=""),
    fresh: bool = Query(default=False),
    caller: CallerIdentity = Depends(require_caller),
) -> dict[str, Any]:
    """Run a read-only Reliability / Availability diagnostic for the configured
    resources and return severity-ranked findings.

    `fresh=true` bypasses the 30 s result cache (used by the page's "Re-run").
    An unknown category is 404; a missing subscription is 400 (the SPA renders
    "open the Setup Wizard first" before calling).
    """
    if category not in supported_categories():
        raise HTTPException(404, f"unknown diagnostic category: {category}")
    sub = subscription_id or _sub_default()
    if not sub:
        raise HTTPException(400, "subscription_id required")

    from api.routes import monitor as monitor_package

    target = DiagnosticTarget(
        subscription_id=sub,
        workload_resource_group=workload_resource_group,
        acr_resource_group=acr_resource_group,
        acr_name=acr_name,
        storage_account_name=storage_account_name,
        region=region,
    )
    del caller  # auth enforced by dependency; identity not needed here.
    return run_diagnostic(category, monitor_package.get_credential(), target, fresh=fresh)

File: web/src/api/diagnostics.ts
/**
 * diagnostics — typed client for `GET /api/diagnostics/{category}`.
 *
 * Read-only Reliability / Availability best-practice findings over the
 * configured Azure resources. Unlike the monitor endpoints, this surface does
 * not degrade open: a fetch failure / permission denial arrives as an
 * `indeterminate` finding, so "could not check" is never rendered as "ok".
 *
 * The `Severity` / `DiagnosticCategory` unions mirror the backend
 * `api/services/diagnostics/models.py`. They are intentionally open at the edge
 * — `severityRank` falls back for an unknown value so a newer backend that adds
 * a severity does not break an older SPA.
 */
import { api } from "@/api/client";

export type Severity = "ok" | "info" | "warning" | "critical" | "indeterminate";
export type DiagnosticCategory = "reliability" | "availability";
export type ResourceKind =
  | "aks"
  | "storage"
  | "acr"
  | "container_app"
  | "api"
  | "queue";

export interface Finding {
  id: string;
  category: DiagnosticCategory;
  pillar: string;
  resource_kind: ResourceKind;
  resource_name: string;
  severity: Severity;
  title: string;
  detail: string;
  recommendation: string;
  doc_url: string;
  rule_version: string;
  expected_by_charter: boolean;
  observed: Record<string, string>;
}

export interface DiagnosticReport {
  category: DiagnosticCategory;
  generated_at: string;
  findings: Finding[];
  rollup: Record<string, number>;
  has_indeterminate: boolean;
}

const SEVERITY_ORDER: Record<string, number> = {
  critical: 4,
  indeterminate: 3,
  warning: 2,
  info: 1,
  ok: 0,
};

/** Sort key for a severity string (unknown values sort lowest). */
export function severityRank(severity: string): number {
  return SEVERITY_ORDER[severity] ?? -1;
}

export interface DiagnosticTargetParams {
  subscriptionId: string;
  workloadResourceGroup?: string;
  acrResourceGroup?: string;
  acrName?: string;
  storageAccountName?: string;
  region?: string;
}

function querystring(params: Record<string, string>): string {
  const usp = new URLSearchParams();
  for (const [k, v] of Object.entries(params)) {
    if (v) usp.set(k, v);
  }
  return usp.toString();
}

export const diagnosticsApi = {
  report: (
    category: DiagnosticCategory,
    target: DiagnosticTargetParams,
    fresh = false,
  ) => {
    const qs = querystring({
      subscription_id: target.subscriptionId,
      workload_resource_group: target.workloadResourceGroup ?? "",
      acr_resource_group: target.acrResourceGroup ?? "",
      acr_name: target.acrName ?? "",
      storage_account_name: target.storageAccountName ?? "",
      region: target.region ?? "",
      fresh: fresh ? "true" : "",
    });
    return api.get<DiagnosticReport>(`/diagnostics/${category}?${qs}`);
  },
};
